CRM/Contribute/BAO/Contribution.php

```diff
@@ -5384,68 +5381,84 @@ protected static function isPaymentInstrumentChange(&$params, $pendingStatuses)
    * load them in this function. Code clean up would compensate for any minor performance implication.
    *
    * @param \CRM_Contribute_BAO_Contribution $contribution
-   * @param array $memberships
    * @param int $primaryContributionID
    * @param string $changeDate
-   * @param string $contributionStatus
-   *   This shouldn't be required but historical function overload by repeattransaction probably requires it.
    *
-   * @todo investigate completely bypassing this function if $contributionStatus != Completed.
    */
-  protected static function updateMembershipBasedOnCompletionOfContribution($contribution, $memberships, $primaryContributionID, $changeDate, $contributionStatus = 'Completed') {
-    foreach ($memberships as $membershipTypeIdKey => $membership) {
+  protected static function updateMembershipBasedOnCompletionOfContribution($contribution, $primaryContributionID, $changeDate) {
+    // Load memberships from contribution info. There may be multiple memberships (or none)
+    $memberships = array();
+    if (!empty($contribution->contribution_recur_id)) {
+      // Load memberships associated with recurring contribution
+      $membershipResult = civicrm_api3('Membership', 'get', array(
+        'contribution_recur_id' => $contribution->contribution_recur_id,
+      ));
+    }
+    elseif (!empty($primaryContributionID)) {
+      // Load membership associated with original contribution
+      $membershipPaymentResult = civicrm_api3('MembershipPayment', 'get', array(
+        'contribution_id' => $primaryContributionID,
+      ));
+      if (!empty($membershipPaymentResult['count'])) {
+        foreach ($membershipPaymentResult['values'] as $payment) {
+          $membershipIDs[] = $payment['membership_id'];
+        }
+        $membershipResult = civicrm_api3('Membership', 'get', array(
```

Comment:
hmmm, why not just get the memberships using $primaryContributionID directly instead of fetching the payments IDs from the payments  first ?

Reply:
@omarabuhussein We are working with a single contribution at this point.  Membership.get requires a recurring contribution ID which we don't have.  Am I missing something?